src/Microsoft.Azure.CosmosEventSourcing/ChangeFeed/EventSourcingProcessor.cs

```diff
@@ -0,0 +1,101 @@
+    private readonly ILogger<EventSourcingProcessor<TSourcedEvent>> _logger;
+    private readonly ISourceProjectionBuilder<TSourcedEvent> _projectionBuilder;
+    private readonly IServiceProvider _serviceProvider;
+    private ChangeFeedProcessor? _processor;
+
+    public EventSourcingProcessor(
+        EventSourcingProcessorOptions<TSourcedEvent> options,
+        ICosmosContainerService containerService,
+        ILeaseContainerProvider leaseContainerProvider,
+        ILogger<EventSourcingProcessor<TSourcedEvent>> logger,
+        ISourceProjectionBuilder<TSourcedEvent> projectionBuilder,
+        IServiceProvider serviceProvider)
+    {
+        _options = options;
+        _containerService = containerService;
+        _leaseContainerProvider = leaseContainerProvider;
+        _logger = logger;
+        _projectionBuilder = projectionBuilder;
+        _serviceProvider = serviceProvider;
+    }
+
+    public async Task StartAsync()
+    {
+        Container itemContainer = await _containerService.GetContainerAsync<TSourcedEvent>();
+        Container leaseContainer = await _leaseContainerProvider.GetLeaseContainerAsync();
+
+        ChangeFeedProcessorBuilder builder = itemContainer
+            .GetChangeFeedProcessorBuilder<TSourcedEvent>(_options.ProcessorName,
+                (changes, token) => OnChangesAsync(changes, token, itemContainer.Id))
+            .WithLeaseContainer(leaseContainer)
+            .WithInstanceName(_options.InstanceName)
+            .WithErrorNotification((token, exception) => OnErrorAsync(exception, itemContainer.Id));
+
+        if (_options.PollInterval.HasValue)
+        {
+            builder.WithPollInterval(_options.PollInterval.Value);
+        }
+
+        _processor = builder.Build();
+
+        _logger.LogInformation("Starting change feed processor for container {ContainerName}", itemContainer.Id);
+
+        await _processor.StartAsync();
+
+        _logger.LogInformation("Successfully started change feed processor for container {ContainerName}",
+            itemContainer.Id);
+    }
+
+    private async Task OnChangesAsync(
+        IReadOnlyCollection<TSourcedEvent> changes,
+        CancellationToken cancellationToken,
+        string containerName)
+    {
+        _logger.LogDebug("Detected changes for container {ContainerName} total ({ChangesCount})",
+            containerName, changes.Count);
+
+        foreach (TSourcedEvent change in changes)
+        {
+            try
+            {
+                await _projectionBuilder.ProjectAsync(change, cancellationToken);
+            }
+            catch (Exception e)
+            {
+                _logger.LogError(e,
+                    "Failed handling projection for container {ContainerName} source event ID {SourcedEventId}",
+                    containerName, change.Id);
+            }
+        }
+    }
+
+    private Task OnErrorAsync(Exception exception, string containerName)
```

Comment:
Should we also broadcast errors for consumers, who might want (or need) to take a different action if something goes wrong?

Reply:
Yes, I like that idea. I am not sure what triggers this as there is no reference to a specific change. I will need to read the docs.